src/asset_manager/common.rs

```diff
@@ -0,0 +1,239 @@
+        DirectoryStore { path: path.as_ref().into() }
+    }
+}
+
+impl AssetStore for DefaultStore {
+    fn read_asset<T, F>(&self, name: &str, format: &F) -> Result<Box<[u8]>, AssetStoreError>
+        where F: AssetFormat,
+              T: Asset
+    {
+        read_asset::<T, _>(name, format)
+    }
+}
+
+#[cfg(not(android))]
+impl AssetStore for DirectoryStore {
+    fn read_asset<T, F>(&self, name: &str, format: &F) -> Result<Box<[u8]>, AssetStoreError>
+        where F: AssetFormat,
+              T: Asset
+    {
+        let path = self.path.join(T::category()).join(name);
+
+        read_asset_from_path(&path, format)
+    }
+}
+
+impl ZipStore {
+    /// Reads a zip file from a stream
+    /// and returns a ZipStore.
+    pub fn from<T>(stream: T) -> Result<Self, IoError>
+        where T: Read + Seek
+    {
+        use zip::read::ZipFile;
+        use zip::result::ZipError;
+
+        let map_zip_err = |x: ZipError| {
+            let e: IoError = match x {
+                    ZipError::Io(x) => x.kind(),
+                    ZipError::FileNotFound => ErrorKind::NotFound,
+                    ZipError::InvalidArchive(_) => ErrorKind::InvalidData,
+                    ZipError::UnsupportedArchive(_) => ErrorKind::InvalidData,
+                }
+                .into();
+
+            e
+        };
+
+        let archive: Result<ZipArchive<T>, ZipError> = ZipArchive::<T>::new(stream);
+        let mut archive = archive.map_err(|x| map_zip_err(x))?;
+
+        let mut map = HashMap::with_capacity(archive.len());
+
+        for file in 0..archive.len() {
+            let mut file: ZipFile = archive.by_index(file).map_err(|x| map_zip_err(x))?;
+            let mut bytes = Vec::with_capacity(file.size() as usize);
+            file.read_to_end(&mut bytes)?;
+            map.insert(file.name().to_owned(), bytes.into_boxed_slice());
+        }
+
+        Ok(ZipStore { files: map })
+    }
+
+    /// Open a file and read its contents
+    #[cfg(not(android))]
+    pub fn open<P: AsRef<Path>>(p: P) -> Result<Self, IoError> {
+        let stream = BufReader::new(File::open(p)?);
+
+        Self::from(stream)
+    }
+
+    /// Open a zip file and read its contents
+    #[cfg(android)]
+    pub fn open<P: AsRef<Path>>(p: P) -> Result<Self, IoError> {
+        use std::io::Cursor;
+
+        let stream = read_bytes(p.to_str().unwrap())?;
+        let stream = Cursor::new(stream);
+
+        Self::from(stream)
+    }
+}
+
+impl AssetStore for ZipStore {
+    fn read_asset<T, F>(&self, name: &str, _: &F) -> Result<Box<[u8]>, AssetStoreError>
+        where F: AssetFormat,
+              T: Asset
+    {
+        let mut pathbuf = PathBuf::from(T::category()).join(name);
+
+        for extension in F::file_extensions() {
+            pathbuf.set_extension(extension);
+            let s = pathbuf.to_str().unwrap();
+
+            if let Some(bytes) = self.files.get(s) {
+                return Ok(bytes.clone());
```

Comment:
Is there a way this could be optimized out if `import` does not consume the bytes?